### DISCUSSION TITLE
linux how to use spdlog in multi-library？

### DISCUSSION THREAD

**Discussion by gakki2019** [gabime/spdlog#3208] [Q&A] [answered]
Linux platform: There are several dynamic library(so), for example:  libmy1.so, libmy2.so, ...
All of those libraries use spdlog to record loggers in console and **same logger file.**
I have read wiki (https://github.com/gabime/spdlog/wiki/How-to-use-spdlog-in-DLLs), but I am still confused how to implement.

libmy1.so, libmy2.so have **no** such API like "setup_logger" to setup sinks.
I put spdlog into a singleton class,  

**logger.hpp**
```
class logger {
public:
    static spdlog::logger *get_instance() {
        static logger instance;
        return instance.m_logger.get();
    }

    void set_level(spdlog::level::level_enum lv);

private:
    logger() noexcept;
    logger(const logger &) = delete;
    logger &operator=(const logger &) = delete;

private:
    std::shared_ptr<spdlog::logger> m_logger;
};
```
**logger.cpp**
```
logger::logger() noexcept {
    auto console_sink = std::make_shared<spdlog::sinks::stdout_color_sink_mt>();
    console_sink->set_level(spdlog::level::trace);
    console_sink->set_pattern("[%Y-%m-%d %H:%M:%S.%e][%t][%l]%v");

    auto file_sink = std::make_shared<spdlog::sinks::rotating_file_sink_mt>("log/log.txt", 1024*1024, 5);
    file_sink->set_level(spdlog::level::trace);
    file_sink->set_pattern("[%Y-%m-%d %H:%M:%S.%e][%t][%l]%v");

    m_logger = std::make_shared<spdlog::logger>(spdlog::logger{"axcl_logger", {console_sink, file_sink}});
    spdlog::register_logger(m_logger);
    m_logger->set_level(spdlog::level::debug);
    m_logger->flush_on(spdlog::level::err);
}
```

Also provide macro in logger.hpp  like：
```
#define MY_LOGGER logger::get_instance()
#define LOG_M_D(tag, fmt, ...)   MY_LOGGER ->log(spdlog::level::debug,    "[{}]: " fmt, tag, ##__VA_ARGS__)
#define LOG_M_I(tag, fmt, ...)     MY_LOGGER ->log(spdlog::level::info,        "[{}]: " fmt, tag, ##__VA_ARGS__)
#define LOG_M_W(tag, fmt, ...)   MY_LOGGER ->log(spdlog::level::warn,     "[{}]: " fmt, tag, ##__VA_ARGS__)
```

All libraries linked to logger.hpp, and logger.cpp, such as:

log
    -   logger.hpp
    -    logger.cpp
   
 libmy1
    - my1fun.cpp
    - Makefile                     #  gcc my1fun.cpp  **log/logger.cpp** -shared -fPIC  libmy1.so     
    
 libmy2
    - my2func.cpp
    - Makefile                     #  gcc my2fun.cpp  **log/logger.cpp** -shared -fPIC  libmy2.so
    
   

Libraries print logger like：
```
#include "log/logger.hpp"
void function_in_my1_library() {
     LOG_M_I("tag", "hello {} in libmy1.so". "spdlog");
}
```

```
#include "log/logger.hpp"
void function_in_my2_library() {
     LOG_M_I("tag", "hello {}" in libmy2.so. "spdlog");
}
```

@tt4g   could you help me? Thank you so much.

**Comment by tt4g** [comment]
I don't know what exactly you are having trouble doing.

At least as previously answered, function bodies with singleton variables should not be listed in header files: https://github.com/gabime/spdlog/discussions/3171#discussioncomment-10601799
And template functions can be used in C++, but not in C (#3173).

**Reply by gakki2019** [reply]
Thanks for reply. 
[#3171](https://github.com/gabime/spdlog/discussions/3171#discussioncomment-10601799) has fixed, there is no issue.
[#3173](https://github.com/gabime/spdlog/discussions/3173) , yes, I have got your point.


In [wiki] (https://github.com/gabime/spdlog/wiki/How-to-use-spdlog-in-DLLs) shows how to use in DLLs, but it seems setup_logger API should be exported by the DLLs, which is not what I expected.

I’m still unsure how to use spdlog effectively across multiple libraries. Here's what I have done so far:
     I provided a common logger.hpp and logger.cpp, which all the libraries compile and include to log into a single file.
     Each library is compiled with the -fvisibility=hidden option to hide unnecessary symbols.
     However, this approach led to a few issues:
    1. Each library ends up with its own copy of the logger instance because the logger class is not marked with __attribute__((visibility("default"))).
    2.  It's difficult to change the log level for each library because each logger class has a own copy.
    3. The logging performance appears to be slow. I’m not sure if this is related to I/O and I haven't tried spdlog's async mode to improve efficiency (I will do that)
   
To summarize:
      There are several libraries on a Linux platform, and all of these libraries need to log into the same file using spdlog.
Could you guide me on how to implement this correctly? If there is a sample, I would greatly appreciate it.

**Reply by tt4g** [reply]
Did you use spdlog as a header-only library in each library? If so, each library will generate its own copy of spdlog.
If you want all libraries to share one spdlog, then spdlog must be built as a shared library.

**Reply by gakki2019** [reply]
Thanks to remind. 
I did use spdlog as head-only.

Now I built spdlog to a shared libary by cmake: libspdlog.so, libspdlog.so1.14, libspdlog.so.1.14.1

All libraries have linked to libspdlog.so like (Makefile) :
CLIB                      += -L$(SPDLOG_LIB_PATH) -lspdlog

Then modify logger.cpp to add _m_logger = spdlog::get("my_logger");_

```
logger::logger() noexcept {
    m_logger = spdlog::get("my_logger");
    if (!m_logger) {
        printf("my_logger++++++++++++++++++++++++++++++++++++++++++++\n");
        auto console_sink = std::make_shared<spdlog::sinks::stdout_color_sink_mt>();
        console_sink->set_level(spdlog::level::trace);
        console_sink->set_pattern("[%Y-%m-%d %H:%M:%S.%e][%t][%L]%v");

        auto file_sink = std::make_shared<spdlog::sinks::rotating_file_sink_mt>("logs/axcl_logs.txt", 5 * 1024 * 1024, 5);
        file_sink->set_level(spdlog::level::trace);
        file_sink->set_pattern("[%Y-%m-%d %H:%M:%S.%e][%t][%L]%v");

        m_logger = std::make_shared<spdlog::logger>(spdlog::logger{"my_logger", {console_sink, file_sink}});
        spdlog::register_logger(m_logger);
        m_logger->set_level(spdlog::level::debug);
        m_logger->flush_on(spdlog::level::err);
    }
}
```

After launch program,  each library will print "my_logger++++++++++++++++++++++++++++++++++++++++++++"

```
my_logger++++++++++++++++++++++++++++++++++++++++++++
my_logger++++++++++++++++++++++++++++++++++++++++++++
[2024-10-08 22:50:48.637][621][info][context manager][create_context][39]: context 1 is created, bind to thread 621
[2024-10-08 22:50:48.638][621][info][device manager][init][45]: device manager is initialized
[2024-10-08 22:50:48.638][623][info][device manager][listen][235]: +++
device id: 129
[2024-10-08 22:50:48.638][623][info][device manager][listen][246]: select +++
[2024-10-08 22:50:48.638][621][info][device][open][33]: 129 +++
my_logger++++++++++++++++++++++++++++++++++++++++++++
```

It seems spdlog registry not work in my code.

**Reply by tt4g** [reply, marked as answer]
This is because your library is still using header-only spdlog.
You must define the `SPDLOG_COMPILED_LIB` macro to use the shared library spdlog (duplicate #1470, #2494).
